File: apps/ios-starter/ThtwaatStarter/Core/Network/SSEClient.swift
import Foundation

struct SSEEvent: Sendable {
    let event: String
    let data: String

    var text: String? {
        guard let data = data.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return data.isEmpty ? nil : data
        }
        return (obj["text"] as? String) ?? (obj["reply"] as? String) ?? (obj["token"] as? String)
    }
}

actor SSEClient {
    private let baseURL: URL
    private let sessionStore: any SessionStore
    private let session: URLSession

    init(baseURL: URL, sessionStore: any SessionStore, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.sessionStore = sessionStore
        self.session = session
    }

    func streamPost(path: String, jsonBody: Data) -> AsyncThrowingStream<SSEEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = URLRequest(url: baseURL.appendingPathComponent(path.hasPrefix("/") ? String(path.dropFirst()) : path))
                    request.httpMethod = "POST"
                    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.httpBody = jsonBody
                    let access = await sessionStore.accessToken()
                    let apiKey = await sessionStore.apiKey()
                    let bearer = access ?? apiKey
                    if let bearer, !bearer.isEmpty {
                        request.setValue("Bearer \(bearer)", forHTTPHeaderField: "Authorization")
                    }
                    if let apiKey, !apiKey.isEmpty {
                        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
                    }

                    let (bytes, response) = try await session.bytes(for: request)
                    if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                        throw APIError(message: "SSE failed: \(http.statusCode)", status: http.statusCode, code: "http_error")
                    }

                    var event = "message"
                    var dataBuf = ""
                    for try await line in bytes.lines {
                        if Task.isCancelled { break }
                        if line.isEmpty {
                            if !dataBuf.isEmpty {
                                continuation.yield(SSEEvent(event: event, data: dataBuf))
                                dataBuf = ""
                                event = "message"
                            }
                            continue
                        }
                        if line.hasPrefix(":") { continue }
                        if line.hasPrefix("event:") {
                            event = line.dropFirst(6).trimmingCharacters(in: .whitespaces)
                        } else if line.hasPrefix("data:") {
                            let chunk = line.dropFirst(5).trimmingCharacters(in: .whitespaces)
                            if !dataBuf.isEmpty { dataBuf += "\n" }
                            dataBuf += chunk
                        }
                    }
                    if !dataBuf.isEmpty {
                        continuation.yield(SSEEvent(event: event, data: dataBuf))
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: apps/android-starter/app/src/main/java/com/thtwaat/starter/core/network/SseClient.kt
package com.thtwaat.starter.core.network

import com.thtwaat.starter.BuildConfig
import com.thtwaat.starter.core.datastore.SessionStore
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.isActive
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.coroutineContext

data class SseEvent(val event: String, val data: String) {
    fun asJson(json: Json = Json { ignoreUnknownKeys = true }): JsonObject? = try {
        json.parseToJsonElement(data).jsonObject
    } catch (_: Exception) {
        null
    }
}

@Singleton
class SseClient @Inject constructor(
    private val okHttpClient: OkHttpClient,
    private val sessionStore: SessionStore,
) {
    fun streamPost(path: String, jsonBody: String): Flow<SseEvent> = flow {
        val token = sessionStore.getAccessToken()
        val apiKey = sessionStore.getApiKey()
        val url = BuildConfig.API_BASE_URL.trimEnd('/') + path
        val builder = Request.Builder()
            .url(url)
            .post(jsonBody.toRequestBody("application/json".toMediaType()))
            .header("Accept", "text/event-stream")
            .header("Content-Type", "application/json")
        val bearer = token ?: apiKey
        if (!bearer.isNullOrBlank()) builder.header("Authorization", "Bearer $bearer")
        if (!apiKey.isNullOrBlank()) builder.header("X-API-Key", apiKey)

        okHttpClient.newCall(builder.build()).execute().use { response ->
            if (!response.isSuccessful) {
                throw ApiException("SSE failed: ${response.code}", status = response.code)
            }
            val source = response.body?.source() ?: return@use
            var event = "message"
            val data = StringBuilder()
            while (coroutineContext.isActive && !source.exhausted()) {
                val line = source.readUtf8Line() ?: break
                when {
                    line.isEmpty() -> {
                        if (data.isNotEmpty()) {
                            emit(SseEvent(event, data.toString()))
                            data.clear()
                            event = "message"
                        }
                    }
                    line.startsWith(":") -> Unit
                    line.startsWith("event:") -> event = line.removePrefix("event:").trim()
                    line.startsWith("data:") -> {
                        if (data.isNotEmpty()) data.append('\n')
                        data.append(line.removePrefix("data:").trimStart())
                    }
                }
            }
            if (data.isNotEmpty()) emit(SseEvent(event, data.toString()))
        }
    }.flowOn(Dispatchers.IO)

    fun extractText(event: SseEvent): String? {
        val obj = event.asJson() ?: return null
        return obj["text"]?.jsonPrimitive?.content
            ?: obj["reply"]?.jsonPrimitive?.content
            ?: obj["token"]?.jsonPrimitive?.content
    }
}
